### PR TITLE
[Core] Optimize `LineNearestPoint` and `TriangleNearestPoint`

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Optimize `LineNearestPoint` and `TriangleNearestPoint` avoiding creating lines (geometry creation is expensive), and computing directly from two points that define the line.

Also adding new implementation based on https://www.shadertoy.com/view/ttfGWl, which is faster and simpler.

Adding scenarios methods requested by @pooyan-dadvand 

**🆕 Changelog**

- [Optimize `LineNearestPoint` and `TriangleNearestPoint`](https://github.com/KratosMultiphysics/Kratos/commit/d6280e15bb35f94b96424eb556595d9f1ad3f568)
- [Implement alternative faster version of TriangleNearestPoint](https://github.com/KratosMultiphysics/Kratos/pull/12586/commits/5022fcf42036dbaebcf5f287302cc1a0793f02f8)
- [Adding test for new implementation](https://github.com/KratosMultiphysics/Kratos/pull/12586/commits/e5a5bd8c8d979dc4e756165919fd9555bd432763)
- [Adding methods with scenarios](https://github.com/KratosMultiphysics/Kratos/pull/12586/commits/2888fbf5a789624fb1d25cfb7ad650abc9ec41d0)
- [Adding tests for scenarios](https://github.com/KratosMultiphysics/Kratos/pull/12586/commits/956a952e58163278de9f4e0a76c42ffcfd83734c)
- Changes from https://github.com/KratosMultiphysics/Kratos/pull/12594
- Changes from https://github.com/KratosMultiphysics/Kratos/pull/12595
- Changes from https://github.com/KratosMultiphysics/Kratos/pull/12598

### CODE REVIEW

**Comment by loumalouomega** [comment]
I removed the computation of some sqrt to improve performance

**Comment by loumalouomega** [comment]
> I removed the computation of some sqrt to improve performance

Where not actually removed, with square distances is enougth

File: kratos/utilities/geometry_utilities/nearest_point_utilities.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+    // Compute square_length of the projection (B)
+    lx = result[0] - rLinePointB[0];
+    ly = result[1] - rLinePointB[1];
+    lz = result[2] - rLinePointB[2];
+    double square_length_2 = lx * lx + ly * ly + lz * lz;
+
+    // Project point locally into the line
+    array_1d<double,3> projected_local;
+    const double tolerance = 1e-14; // Tolerance
```
Comment (pooyan-dadvand) [inline kratos/utilities/geometry_utilities/nearest_point_utilities.cpp]: I would take the tolerance as an optional argument.

File: kratos/utilities/geometry_utilities/nearest_point_utilities.cpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    // Project point globally into the line
+    const array_1d<double, 3> ab = rLinePointB - rLinePointA;
+    const double factor = (inner_prod(rLinePointB, rPoint) - inner_prod(rLinePointA, rPoint) - inner_prod(rLinePointB, rLinePointA) + inner_prod(rLinePointA, rLinePointA)) / inner_prod(ab, ab);
+    Point result(rLinePointA + factor * ab);
+
+    // Compute square_length of the line
+    double lx = rLinePointA[0] - rLinePointB[0];
+    double ly = rLinePointA[1] - rLinePointB[1];
+    double lz = rLinePointA[2] - rLinePointB[2];
+    double square_length = lx * lx + ly * ly + lz * lz;
```
Comment (pooyan-dadvand) [inline kratos/utilities/geometry_utilities/nearest_point_utilities.cpp]: const double

File: kratos/utilities/geometry_utilities/nearest_point_utilities.cpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+    // Project point locally into the line
+    array_1d<double,3> projected_local;
+    const double tolerance = 1e-14; // Tolerance
+
+    if (square_length_1 <= (square_length + tolerance) && square_length_2 <= (square_length + tolerance)) {
+        return result;
+    } else {
+        // If the projected point is outside the line, return the nearest end point
+        array_1d<double, 3> auxiliary_values = rLinePointA - result;
```
Comment (pooyan-dadvand) [inline kratos/utilities/geometry_utilities/nearest_point_utilities.cpp]: Are you sure that you need to calculate again the distances? We cannot reuse the square lengths to decide which point?

File: kratos/utilities/geometry_utilities/nearest_point_utilities.h [hunk 4]
```diff
@@ -74,134 +97,160 @@ class NearestPointUtilities
     ///@{
 
     /**
-     * @brief Finds the nearest point to the given point on a line segment. 
-     * @details It first projects the point into the line. If the porjceted point is inside the segment boundary 
+     * @brief Finds the nearest point to the given point on a line segment.
+     * @details It first projects the point into the line. If the projected point is inside the segment boundary
```
Comment (pooyan-dadvand) [inline kratos/utilities/geometry_utilities/nearest_point_utilities.h:101]: I would add the reference of the algorithm

File: kratos/utilities/geometry_utilities/nearest_point_utilities.cpp (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities/nearest_point_utilities.cpp]: Okay

File: kratos/utilities/geometry_utilities/nearest_point_utilities.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities/nearest_point_utilities.h:101]: Okay

File: kratos/utilities/geometry_utilities/nearest_point_utilities.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/geometry_utilities/nearest_point_utilities.cpp]: Okay

Comment (loumalouomega) [reply kratos/utilities/geometry_utilities/nearest_point_utilities.cpp]: Let me check

Comment (loumalouomega) [reply kratos/utilities/geometry_utilities/nearest_point_utilities.cpp]: You are right (in fact I calculated twice the vey same)

**Review by pooyan-dadvand — APPROVED** [review]
Thanks!

**Pull request merged**